flasc/analysis/expected_power_analysis_utilities.py

```diff
@@ -0,0 +1,484 @@
+            for col in sync_cols
+        ]
+
+        df_ = joined_df.with_columns(updated_columns).select(df_.columns)  # Return original columns
+
+    return df_
+
+
+def _get_num_points_pair(
+    df_: pl.DataFrame,
+    test_cols: List[str],
+    bin_cols_with_df_name: List[str],
+) -> pl.DataFrame:
+    """Get the number of points for each pair of test columns.
+
+    Args:
+        df_ (pl.DataFrame): A polars dataframe
+        test_cols (List[str]): A list of column names to calculate the number of points for
+        bin_cols_with_df_name (List[str]): A list of column names to bin the dataframes by.
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the number of points for each pair of test columns.
+    """
+    # Generate all pairs of columns (including same column pairs)
+    col_pairs = list(product(test_cols, test_cols))
+
+    df_n = df_.group_by(bin_cols_with_df_name).agg(pl.len().alias("count"))
+
+    for c1, c2 in col_pairs:
+        df_sub = df_.filter(
+            pl.col(c1).is_not_null()
+            & pl.col(c2).is_not_null()
+            & pl.col(c1).is_not_nan()
+            & pl.col(c2).is_not_nan()
+        )
+        df_n_sub = df_sub.group_by(bin_cols_with_df_name).agg(pl.len().alias(f"count_{c1}_{c2}"))
+
+        df_n = df_n.join(df_n_sub, on=bin_cols_with_df_name, how="left")
+
+    # Fill nulls to 0
+    df_n = df_n.fill_null(0)
+
+    # Sort by bin_cols_with_df_name
+    df_n = df_n.sort(bin_cols_with_df_name)
+
+    return df_n
+
+
+def _compute_covariance(
+    df_: pl.DataFrame,
+    test_cols: List[str],
+    bin_cols_with_df_name: List[str],
+) -> pl.DataFrame:
+    """Compute the covariance matrix for the test columns.
+
+    Args:
+        df_ (pl.DataFrame): A polars dataframe
+        test_cols (List[str]): A list of column names to calculate the covariance of
+        bin_cols_with_df_name (List[str]): A list of column names to bin the dataframes by.
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the covariance matrix.
+    """
+    # Generate all pairs of columns (including same column pairs)
+    col_pairs = list(product(test_cols, test_cols))
+
+    # Create expressions for all covariances
+    cov_exprs = [pl.cov(pair[0], pair[1]).alias(f"cov_{pair[0]}_{pair[1]}") for pair in col_pairs]
+
+    # Compute covariances
+    df_cov = df_.group_by(bin_cols_with_df_name).agg(cov_exprs)
+
+    # Get the number of points for each pair
+    df_n = _get_num_points_pair(
+        df_, test_cols=test_cols, bin_cols_with_df_name=bin_cols_with_df_name
+    )
+
+    # Join the number of points to the covariance matrix
+    df_cov = df_cov.join(df_n, on=bin_cols_with_df_name, how="left")
+
+    # Enforce that each ws/wd bin combination has to appear in all dataframes
+    bin_cols_without_df_name = [c for c in bin_cols_with_df_name if c != "df_name"]
+    num_df = df_["df_name"].n_unique()
+    df_cov = df_cov.filter(pl.len().over(bin_cols_without_df_name) == num_df)
+
```

Comment:
Reminder to set count values of 0 to Null in this function.

Reply:
Realizing it's a bit back and forth since in the function above, _get_num_points_pair, I fill nulls with zeros, so just going the other way and setting zeros to null there

Reply:
Agreed, keeping counts as the actual number of overlapping samples should work fine